mslib/thermolib.py

```diff
@@ -210,141 +211,126 @@ def omega_to_w(omega, p, t):
+            p[indices] = p0 * ((t0 - gamma * (height[indices] - z0)) / t0) ** (g / (gamma * Rd))
         else:
-            p[indices] = p0 * numpy.exp(-g * (z[indices] - z0) / (Rd * t0))
+            p[indices] = p0 * numpy.exp(-g * (height[indices] - z0) / (Rd * t0))
 
     if numpy.isnan(p).any():
         raise ValueError("flight level to pressure conversion not "
                          "implemented for z > 71km")
 
-    return p
-
-
-def pressure2flightlevel_a(p):
-    """
-    Conversion of pressure (Pa) to flight level (hft) with
-    hydrostatic equation, according to the profile of the ICAO
-    standard atmosphere.
+    return p if is_array else p[0]
 
-    Array version, the argument "p" must be a numpy array.
 
+@exporter.export
+@preprocess_and_wrap(wrap_like='pressure')
+@check_units('[pressure]')
+def pressure2flightlevel(pressure):
+    r"""
+    Conversion of pressure to height (hft) with
+    hydrostatic equation, according to the profile of the 1976 U.S. Standard atmosphere [NOAA1976]_.
     Reference:
-        For example, H. Kraus, Die Atmosphaere der Erde, Springer, 2001,
-        470pp., Sections II.1.4. and II.6.1.2.
-
-    Arguments:
-        p -- numpy array of pressure (Pa)
-        fake_above_32km -- compute values above 54.75 hPa (32km) with the
-                           profile valid for 20..32km. WARNING: This gives
-                           unphysical results. Use this option only for
-                           testing purposes.
-    Returns:
-        flight level in hft
+        H. Kraus, Die Atmosphaere der Erde, Springer, 2001, 470pp., Sections II.1.4. and II.6.1.2.
+
+    Parameters
+    ----------
+    pressure : `pint.Quantity` or `xarray.DataArray`
+        Atmospheric pressure
+
+    Returns
+    -------
+    `pint.Quantity` or `xarray.DataArray`
+        Corresponding height value(s) (hft)
+
+    Notes
+    -----
+    .. math:: Z = \begin{cases}
+              Z_0 + \frac{T_0 - T_0 \cdot \exp\left(\frac{\Gamma \cdot R}{g\cdot\log(\frac{p}{p0})}\right)}{\Gamma}
+              &\Gamma \neq 0\\
+              Z_0 - \frac{R \cdot T_0}{g \cdot \log(\frac{p}{p_0})} &\text{else}
+              \end{cases}
     """
-    # Make sure p is a numpy array.
-    if not isinstance(p, numpy.ndarray):
-        raise ValueError("argument p must be a numpy array")
+    is_array = hasattr(pressure.magnitude, "__len__")
+    if not is_array:
+        pressure = [pressure.magnitude] * pressure.units
 
     # Initialize the return array.
-    z = numpy.full_like(p, numpy.nan)
+    z = numpy.full_like(pressure, numpy.nan) * units.hft
 
-    for i, ((z0, t0, gamma, p0), (z1, t1, _, p1)) in enumerate(zip(ZTGPS[:-1], ZTGPS[1:])):
-        p1 = ZTGPS[i + 1][-1]
-        indices = (p > p1) & (p <= p0)
+    for i, ((z0, t0, p0, gamma), (z1, t1, p1, _)) in enumerate(zip(standard_atmosphere[:-1], standard_atmosphere[1:])):
+        p1 = standard_atmosphere[i + 1][-2]
```

Comment:
maybe use named constants instead of -2?

Reply:
Done